src/core/prometheus.cc

```diff
@@ -574,39 +613,28 @@ future<> write_text_representation(output_stream<char>& out, const config& ctx,
                         write_histogram(s, ctx, name, value.get_histogram(), value_info.id.labels());
                     }
                 } else {
-                    add_name(s, name, value_info.id.labels(), ctx);
-                    std::string value_str;
-                    try {
-                        value_str = to_str(value);
-                    } catch (const std::range_error& e) {
-                        seastar_logger.debug("prometheus: write_text_representation: {}: {}", s.str(), e.what());
-                        value_str = "NaN";
-                    } catch (...) {
-                        auto ex = std::current_exception();
-                        // print this error as it's ignored later on by `connection::start_response`
-                        seastar_logger.error("prometheus: write_text_representation: {}: {}", s.str(), ex);
-                        std::rethrow_exception(std::move(ex));
+                    if (should_aggregate) {
+                        counters.add_counter(value, value_info.id.labels());
+                    } else {
+                        write_counter(s, ctx, name, value, value_info.id.labels());
                     }
-                    s << value_str;
-                    s << "\n";
                 }
                 out.write(s.str()).get();
                 thread::maybe_yield();
             });
-            if (!histograms.empty()) {
-                auto name = ctx.prefix + "_" + metric_family.name();
-                std::stringstream name_help;
-                if (metric_family.metadata().d.str() != "") {
-                    name_help << "# HELP " << name << " " <<  metric_family.metadata().d.str() << "\n";
-                }
-                name_help << "# TYPE " << name << " histogram"  << "\n";
-                out.write(name_help.str()).get();
-                for (auto&& h : histograms.get_histograms()) {
-                    std::stringstream s;
-                    write_histogram(s, ctx, name, h.second, h.first);
-                    out.write(s.str()).get();
-                    thread::maybe_yield();
-                }
+
+            for (auto&& h : histograms.get_histograms()) {
+                std::stringstream s;
+                write_histogram(s, ctx, name, h.second, h.first);
+                out.write(s.str()).get();
+                thread::maybe_yield();
+            }
+
+            for (auto&& h : counters.get_counters()) {
+                std::stringstream s;
+                write_counter(s, ctx, name, h.second, h.first);
+                out.write(s.str()).get();
+                thread::maybe_yield();
```

Comment:
It would be more efficient to hoist the `stringstream` from these loops and `.clear()` it each iteration.

Reply:
Oh yeah, I forgot how to use stringstream.  I think you need, `ss.clear(); ss.str("");` - and to be completely correct, a bunch more things, but that should suffice here.